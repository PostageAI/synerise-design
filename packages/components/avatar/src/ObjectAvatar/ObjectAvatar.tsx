import * as React from 'react';
import { MailM } from '@synerise/ds-icon/dist/icons';
import theme from '@synerise/ds-core/dist/js/DSProvider/ThemeProvider/theme';
import Icon from '@synerise/ds-icon';
import Badge from '@synerise/ds-badge';
import Status from '@synerise/ds-status';

import Avatar from '../Avatar';
import { getObjectName, addIconColor, getColorByText } from '../utils';

import { ObjectAvatarProps } from '../Avatar.types';

export const DEFAULT_COLOR_HUE = '600';
export const DEFAULT_COLOR = 'grey';

const ObjectAvatar: React.FC<ObjectAvatarProps> = ({
  backgroundColor = 'grey',
  badgeStatus,
  iconComponent,
  color = 'grey',
  size,
  src,
  tooltip,
  objectName,
  objectDescription,
  objectStatus,
  text,
  children,
  disabled,
  style,
  ...restProps
}) => {
  const defaultTooltip = {
    title: objectName || '',
    description: objectDescription || '',
    status: objectStatus ? <Status label={objectStatus} type="disabled" /> : '',
  };
  const avatarText = getObjectName(objectName, text);
  const [avatarBackgroundColor, avatarBackgroundHue] = getColorByText(avatarText, backgroundColor);
  const iconColor = theme.palette[`${color || DEFAULT_COLOR}-${DEFAULT_COLOR_HUE}`];
  const avatarTooltip =
    tooltip === undefined && (defaultTooltip.title || defaultTooltip.description) ? defaultTooltip : tooltip;

  const iconElement = addIconColor(iconComponent, iconColor);

  const icon = !avatarText ? iconElement || <Icon component={<MailM />} color={iconColor} /> : null;

  const avatar = (
    <Avatar
      iconComponent={icon}
      shape="square"
      backgroundColor={avatarBackgroundColor}
      backgroundColorHue={avatarBackgroundHue}
      size={size}
      hasStatus={disabled === true ? false : !!badgeStatus}
      src={src}
      tooltip={avatarTooltip}
      disabled={disabled}
      {...restProps}
    >
      {children || avatarText}
    </Avatar>
  );

  return badgeStatus ? (
    <span style={style}>
      <Badge status={badgeStatus}>{avatar}</Badge>
    </span>
  ) : (
    avatar
  );
};

export default ObjectAvatar;
