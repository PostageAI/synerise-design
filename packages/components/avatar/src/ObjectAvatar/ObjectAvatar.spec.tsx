import * as React from 'react';
import { renderWithProvider } from '@synerise/ds-utils/dist/testing';
import ObjectAvatar from './ObjectAvatar';

const testObject = {
  name: 'Shovel',
  description: 'Some product',
  status: 'API',
};

const testObjectImage = {
  name: 'Shovel',
  description: 'Some product',
  status: 'API',
  avatar: 'https://image.com/image.jpg',
};

describe('ObjectAvatar', () => {
  test('should render', () => {
    const { container, debug } = renderWithProvider(<ObjectAvatar />);
    debug();
    expect(container.getElementsByClassName('mail-m').length).toBe(1);
  });
  test('should render with object data', () => {
    const objectAvatar = renderWithProvider(<ObjectAvatar object={testObject} />);
    const firstLetters = objectAvatar.getByText(testObject.name.charAt(0));
    expect(firstLetters).toBeTruthy();
  });
  test('should render with object data and avatar', () => {
    const objectAvatar = renderWithProvider(<ObjectAvatar object={testObjectImage} />);
    const img = objectAvatar.getByRole('img');
    expect(img).toHaveAttribute('src', testObjectImage.avatar);
  });
});
