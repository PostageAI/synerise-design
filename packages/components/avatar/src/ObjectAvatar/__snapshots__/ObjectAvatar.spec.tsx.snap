// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ObjectAvatar should match snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <span
        class="ant-avatar Avatarstyles-xcn5o4-1 hiogYD ds-avatar ant-avatar-square"
      >
        <span
          class="ant-avatar-string"
          style="transform: scale(1) translateX(-50%);"
        >
          <div
            class="Iconstyles__IconContainer-aqhvae-0 gYSvHk ds-icon "
            color="#6a7580"
            size="24"
          >
            <svg
              class="mail-m"
              viewBox="0 0 24 24"
            >
              <path
                d="M0 0h24v24H0z"
                fill="none"
                opacity="0.1"
              />
              <path
                d="M17.25 5.75H6.75A2.753 2.753 0 004 8.5V15a2.753 2.753 0 002.75 2.75h10.5A2.753 2.753 0 0020 15V8.5a2.753 2.753 0 00-2.75-2.75zm-10.5 1.5h10.5a1.233 1.233 0 011.054.62l-5.053 3.225a2.265 2.265 0 01-2.409.009L5.696 7.869A1.233 1.233 0 016.75 7.25zm10.5 9H6.75A1.251 1.251 0 015.5 15V9.517l4.544 2.857a3.767 3.767 0 004.014-.014L18.5 9.524V15a1.251 1.251 0 01-1.25 1.25z"
              />
            </svg>
          </div>
        </span>
      </span>
    </div>
  </body>,
  "container": <div>
    <span
      class="ant-avatar Avatarstyles-xcn5o4-1 hiogYD ds-avatar ant-avatar-square"
    >
      <span
        class="ant-avatar-string"
        style="transform: scale(1) translateX(-50%);"
      >
        <div
          class="Iconstyles__IconContainer-aqhvae-0 gYSvHk ds-icon "
          color="#6a7580"
          size="24"
        >
          <svg
            class="mail-m"
            viewBox="0 0 24 24"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
              opacity="0.1"
            />
            <path
              d="M17.25 5.75H6.75A2.753 2.753 0 004 8.5V15a2.753 2.753 0 002.75 2.75h10.5A2.753 2.753 0 0020 15V8.5a2.753 2.753 0 00-2.75-2.75zm-10.5 1.5h10.5a1.233 1.233 0 011.054.62l-5.053 3.225a2.265 2.265 0 01-2.409.009L5.696 7.869A1.233 1.233 0 016.75 7.25zm10.5 9H6.75A1.251 1.251 0 015.5 15V9.517l4.544 2.857a3.767 3.767 0 004.014-.014L18.5 9.524V15a1.251 1.251 0 01-1.25 1.25z"
            />
          </svg>
        </div>
      </span>
    </span>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`ObjectAvatar should match snapshot with object data 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <span
        class="ant-avatar Avatarstyles-xcn5o4-1 hfFXOL ds-avatar ant-avatar-square"
      >
        <span
          class="ant-avatar-string"
          style="transform: scale(1) translateX(-50%);"
        >
          J
        </span>
      </span>
    </div>
  </body>,
  "container": <div>
    <span
      class="ant-avatar Avatarstyles-xcn5o4-1 hfFXOL ds-avatar ant-avatar-square"
    >
      <span
        class="ant-avatar-string"
        style="transform: scale(1) translateX(-50%);"
      >
        J
      </span>
    </span>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
